code/ai/aicode.cpp

```diff
@@ -14636,7 +14642,24 @@ int aas_1(object *objp, ai_info *aip, vec3d *safe_pos)
 			}
 		}
 
-		if (!pos_set) {
+		if (pos_set) {
+			// if the object is not a small ship, a surface impact is likely to be some distance away from the ship center, so refine the position
+			if (fix_bugs && target_ship_obj && !Ship_info[Ships[target_ship_obj->instance].ship_info_index].is_small_ship()) {
+				mc_info mc;
+				mc.model_instance_num = Ships[target_ship_obj->instance].model_instance_num;
+				mc.model_num = Ship_info[Ships[target_ship_obj->instance].ship_info_index].model_num;
+				mc.orient = &target_ship_obj->orient;
+				mc.pos = &target_ship_obj->pos;
+				mc.p0 = &weapon_objp->pos;				// Point 1 of ray to check
+				mc.p1 = &expected_pos;					// Point 2 of ray to check
+				mc.flags = MC_CHECK_MODEL;
+
+				model_collide(&mc);
```

Comment:
I'm loathe to add more collision checks unless absolutely necessary, especially since this appears to be called every frame for a thing which is in the average case unlikely to move much (and also this case doesn't need tons of accuracy, some approximation/lag should be fine), so at the very least the result should probably be cached.

Alternatively, have you considered just using the weapon's `homing_pos`?

Reply:
It's a fair point, and I had the same reluctance, but I ended up adding it after seeing how many collision checks were used elsewhere in the code, and due to the difficulty of solving this problem without a collision check.  I'm certainly open to suggestions though.  And I could cache the result, at the cost of adding some additional fields to `ai_info`.

The weapon's `homing_pos` will work if the missile is homing on a subsystem, but will not work if the weapon is homing on the actual ship, since the position will be at the center of the ship.

Reply:
On a large ship, the `homing_pos` shouldn't be the center of the ship, see the usage of `ai_big_pick_attack_point` in `weapon_home`. They tend to spread themselves over the hull surface on the side facing the missile.

Reply:
Ok, I finally had a chance to get back to this.  I modified the code to print debug output every frame to show the coordinates being used.  Here is the output for homing on one of the turrets:
```
expected_pos: [x=7417.05, y=-108.33, z=637.62]    weapon homing_pos: [x=7421.30, y=-106.31, z=630.08]    center of ship: [x=7862.55, y=0.00, z=581.01]    expected-homing: 8.89    expected-center: 461.97    homing-center: 456.52
expected_pos: [x=7417.05, y=-108.33, z=637.62]    weapon homing_pos: [x=7421.30, y=-106.31, z=630.08]    center of ship: [x=7862.55, y=0.00, z=581.01]    expected-homing: 8.89    expected-center: 461.97    homing-center: 456.52
expected_pos: [x=7417.05, y=-108.33, z=637.62]    weapon homing_pos: [x=7421.30, y=-106.31, z=630.08]    center of ship: [x=7862.55, y=0.00, z=581.01]    expected-homing: 8.89    expected-center: 461.97    homing-center: 456.52
expected_pos: [x=7417.05, y=-108.33, z=637.62]    weapon homing_pos: [x=7421.30, y=-106.31, z=630.08]    center of ship: [x=7862.55, y=0.00, z=581.01]    expected-homing: 8.89    expected-center: 461.97    homing-center: 456.52
expected_pos: [x=7417.05, y=-108.33, z=637.62]    weapon homing_pos: [x=7421.30, y=-106.31, z=630.08]    center of ship: [x=7862.55, y=0.00, z=581.01]    expected-homing: 8.89    expected-center: 461.97    homing-center: 456.52
...
```
And here is the output for homing on the ship with no subsystem selected:
```
expected_pos: [x=7483.91, y=-133.26, z=826.83]    weapon homing_pos: [x=7862.55, y=0.00, z=581.01]    center of ship: [x=7862.55, y=0.00, z=581.01]    expected-homing: 470.70    expected-center: 470.70    homing-center: 0.00
expected_pos: [x=7483.91, y=-133.26, z=826.83]    weapon homing_pos: [x=7862.55, y=0.00, z=581.01]    center of ship: [x=7862.55, y=0.00, z=581.01]    expected-homing: 470.70    expected-center: 470.70    homing-center: 0.00
expected_pos: [x=7483.91, y=-133.26, z=826.83]    weapon homing_pos: [x=7862.55, y=0.00, z=581.01]    center of ship: [x=7862.55, y=0.00, z=581.01]    expected-homing: 470.70    expected-center: 470.70    homing-center: 0.00
expected_pos: [x=7483.91, y=-133.26, z=826.83]    weapon homing_pos: [x=7862.55, y=0.00, z=581.01]    center of ship: [x=7862.55, y=0.00, z=581.01]    expected-homing: 470.70    expected-center: 470.70    homing-center: 0.00
expected_pos: [x=7483.91, y=-133.26, z=826.83]    weapon homing_pos: [x=7862.55, y=0.00, z=581.01]    center of ship: [x=7862.55, y=0.00, z=581.01]    expected-homing: 470.70    expected-center: 470.70    homing-center: 0.00
...
```

This confirms what I discovered last month -- the `homing_pos` is close enough if the weapon is homing on a subsystem (or turret), but is equivalent to the center of the ship if the weapon is homing on the ship.

I took a deeper look through the code and found that the octant picking is never actually called.  This is because the `wp->pick_big_attack_point_timestamp` is always 0 (i.e. invalid) and never elapses.  This is actually intentional:
```C++
// If player has apect lock, we don't want to find a homing point on the closest
// octant... setting the timestamp to 0 ensures this.
if (wip->is_locked_homing()) {
	wp->pick_big_attack_point_timestamp = 0;
} else {
	wp->pick_big_attack_point_timestamp = 1;
}
```
Then the "preliminary" attack point, which is used if the octant picking code fails or is bypassed, is set as follows:
```C++
if (pp && pp->locking_subsys) {
	wp->big_attack_point = pp->locking_subsys->system_info->pnt;
} else {
	vm_vec_zero(&wp->big_attack_point);
}
```

So anyway, that was a bit of a detour, but was informative.  In any case, I'll figure out a way to cache the expected position so the collision check doesn't need to run every frame.

Reply:
Isn't that just for player missiles? (Not that I am trying to imply that is an irrelevant case)

Reply:
Hmm, that's a good point -- that is inside a block that checks for the player ship.